signature-aggregator/aggregator/aggregator.go

```diff
@@ -179,6 +178,115 @@ func (s *SignatureAggregator) connectToQuorumValidators(
 	return connectedValidators, nil
 }
 
+// getUnderfundedL1Nodes fetches the set of L1 nodes that are underfunded
+// It uses the underfundedL1NodeCache to avoid fetching the same data multiple times
+func (s *SignatureAggregator) getUnderfundedL1Nodes(
+	ctx context.Context,
+	log logging.Logger,
+	signingSubnet ids.ID,
+	skipCache bool,
+) (set.Set[ids.NodeID], error) {
+	fetchUnderfundedL1Nodes := func(subnetID ids.ID) (set.Set[ids.NodeID], error) {
+		validators, err := s.pChainClient.GetCurrentValidators(ctx, subnetID, nil, s.pChainClientOptions...)
+		if err != nil {
+			log.Error(
+				"Failed to fetch current L1 validators",
+				zap.String("subnetID", subnetID.String()),
+				zap.Error(err),
+			)
+			return nil, fmt.Errorf("failed to fetch current L1 validators: %w", err)
+		}
+
+		underfundedL1Nodes := set.NewSet[ids.NodeID](0)
+		for _, v := range validators {
+			if v.ClientL1Validator.ValidationID == nil {
+				log.Debug(
+					"Skipping non-L1 validator",
+					zap.String("nodeID", v.NodeID.String()),
+				)
+				continue
+			}
+
+			l1Validator := v.ClientL1Validator
+
+			if l1Validator.Balance == nil {
```

Comment:
Do we know what cases the Balance would be nil? And shouldn't these nodes get added to the underfunded list?